sklearn/utils/_array_transformer.py

```diff
@@ -0,0 +1,88 @@
+import numpy as np
+import scipy.sparse as sp_sparse
+from .._config import get_config
+
+
+def _one_to_one(feature_names_in):
+    return feature_names_in
+
+
+def _get_feature_names(X):
+    if hasattr(X, "columns"):
+        return np.array(X.columns, dtype=object)
+    elif hasattr(X, "dims") and isinstance(X.dims, tuple):
+        # xarray DataArray
+        if len(X.dims) != 2:
+            raise ValueError("XArray.DataArray must be 2D")
+        return np.array(X.coords[X.dims[1]], dtype=object)
+
+
+class _ArrayTransformer:
+
+    def __init__(self, X, needs_feature_names_in=True):
+        if (get_config()['array_out'] == 'default'
+           or not needs_feature_names_in):
+            self.feature_names_in = None
+        else:
+            self.feature_names_in = _get_feature_names(X)
+        self.needs_feature_names_in = needs_feature_names_in
+        self.dims = getattr(X, "dims", None)
+
+    def transform(self, X, get_feature_names_out=_one_to_one):
+        array_out = get_config()['array_out']
+        if array_out == 'default':
+            return X
+
+        if self.needs_feature_names_in:
+            feature_names_out = get_feature_names_out(self.feature_names_in)
+        else:
+            feature_names_out = get_feature_names_out()
+
```

Comment:
Would it be useful to check `len(feature_names_out) == X.shape[1]` rather than land up with the error message in the container construction?